src/OpenApi/src/Extensions/JsonObjectSchemaExtensions.cs

```diff
@@ -0,0 +1,279 @@
+    /// opposed to after the generated schemas have been mapped to OpenAPI schemas.
+    /// </remarks>
+    /// <param name="schema">The <see cref="JsonObject"/> produced by the underlying schema generator.</param>
+    /// <param name="type">The <see cref="Type"/> associated with the <see paramref="schema"/>.</param>
+    internal static void ApplyPrimitiveTypesAndFormats(this JsonObject schema, Type type)
+    {
+        if (_simpleTypeToOpenApiSchema.TryGetValue(type, out var openApiSchema))
+        {
+            schema["nullable"] = openApiSchema.Nullable || (schema["type"] is JsonArray schemaType
+                && schemaType.GetValues<string>().Contains("null"));
+            schema["type"] = openApiSchema.Type;
+            schema["format"] = openApiSchema.Format;
+        }
+    }
+
+    /// <summary>
+    /// Applies route constraints to the target schema.
+    /// </summary>
+    /// <param name="schema">The <see cref="JsonObject"/> produced by the underlying schema generator.</param>
+    /// <param name="constraints">The list of <see cref="IRouteConstraint"/>s associated with the route parameter.</param>
+    internal static void ApplyRouteConstraints(this JsonObject schema, IEnumerable<IRouteConstraint> constraints)
+    {
+        foreach (var constraint in constraints)
+        {
+            if (constraint is MinRouteConstraint minRouteConstraint)
+            {
+                schema["minimum"] = minRouteConstraint.Min;
+            }
+            else if (constraint is MaxRouteConstraint maxRouteConstraint)
+            {
+                schema["maximum"] = maxRouteConstraint.Max;
+            }
+            else if (constraint is MinLengthRouteConstraint minLengthRouteConstraint)
+            {
+                schema["minLength"] = minLengthRouteConstraint.MinLength;
+            }
+            else if (constraint is MaxLengthRouteConstraint maxLengthRouteConstraint)
+            {
+                schema["maxLength"] = maxLengthRouteConstraint.MaxLength;
+            }
+            else if (constraint is RangeRouteConstraint rangeRouteConstraint)
+            {
+                schema["minimum"] = rangeRouteConstraint.Min;
+                schema["maximum"] = rangeRouteConstraint.Max;
+            }
+            else if (constraint is RegexRouteConstraint regexRouteConstraint)
+            {
+                schema["type"] = "string";
+                schema["format"] = null;
+                schema["pattern"] = regexRouteConstraint.Constraint.ToString();
+            }
+            else if (constraint is LengthRouteConstraint lengthRouteConstraint)
+            {
+                schema["minLength"] = lengthRouteConstraint.MinLength;
+                schema["maxLength"] = lengthRouteConstraint.MaxLength;
+            }
+            else if (constraint is FloatRouteConstraint or DecimalRouteConstraint or DoubleRouteConstraint)
```

Comment:
Is there some way we could reuse the values from `_simpleTypeToOpenApiSchema` so they don't get out of sync?

Reply:
There might be although it might end up being more LoC than what we have here...

FWIW, I don't expect the formats to change aggressively and I can add test cases to help us catch if they go out of sync.